### ISSUE TITLE
no-unused-expression is marking `void(0)` as an issue

### ISSUE THREAD

**Issue by ValeriiVasin** [palantir/tslint#2632]
### Bug Report

- __TSLint version__: 5.1.0
- __TypeScript version__: 2.2.2
- __Running TSLint via__: CLI

#### TypeScript code being linted

```ts
const config = {
  callback: (value) => value ? `${value}%` : void(0),
}
```

with `tslint.json` configuration:

```json
{
    "rules": {
        "no-internal-module": true,
        "typedef-whitespace": [
            true,
            {
              "call-signature": "nospace",
              "index-signature": "nospace",
              "parameter": "nospace",
              "property-declaration": "nospace",
              "variable-declaration": "nospace"
            }
        ],

        "curly": true,
        "forin": true,
        "no-bitwise": true,
        "no-arg": true,
        "no-conditional-assignment": true,
        "no-console": [
            true,
            "debug",
            "info",
            "time",
            "timeEnd",
            "trace",
            "log",
            "error"
        ],
        "no-construct": true,
        "no-debugger": true,
        "no-duplicate-variable": true,
        "no-empty": true,
        "no-eval": true,
        "no-invalid-this": true,
        "no-shadowed-variable": true,
        "no-string-literal": true,
        "no-switch-case-fall-through": true,
        "no-unused-expression": true,
        "no-unused-variable": true,
        "no-use-before-declare": true,
        "no-var-keyword": true,
        "radix": true,
        "switch-default": true,
        "triple-equals": true,
        "use-isnan": true,

        "eofline": true,
        "indent": [true, "spaces"],
        "linebreak-style": [true, "LF"],
        "max-line-length": [true, 120],
        "no-trailing-whitespace": true,
        "trailing-comma": [true, {
          "multiline": "always",
          "singleline": "never"
        }],

        "class-name": true,
        "comment-format": [
            true,
            "check-space"
        ],
        "interface-name": [true, "always-prefix"],
        "jsdoc-format": true,
        "no-consecutive-blank-lines": [true, 1],
        "one-line": [
          true,
          "check-catch",
          "check-finally",
          "check-else",
          "check-open-brace",
          "check-whitespace"
        ],
        "quotemark": [true, "single"],
        "semicolon": [true, "always"],
        "variable-name": [
            true,
            "ban-keywords"
        ],
        "whitespace": [
            true,
            "check-branch",
            "check-decl",
            "check-operator",
            "check-module",
            "check-separator",
            "check-type",
            "check-typecast"
        ]
    }
}
```

#### Actual behavior
[tslint] unused expression, expected an assignment or function call (no-unused-expression)

#### Expected behavior
i would assume that all should be fine here

P.S.  If i use `void 0` instead - it is not complaining.

**Comment by ajafff** [comment]
@ValeriiVasin 
The rule checks void expressions the same way it checks statements. That's done to find expressions like `return void foo.bar` or `return void(a==b)` which are technically just unused expressions.
Besides "used expressions" like `return void foo()` or `return void(a = b)` there is just one exception: literal `0` to allow `void 0` as you noted above.

Would it be ok for you to just use `void 0` instead of `void(0)`?

**Comment by ValeriiVasin** [comment]
That is what i did, but i would assume that using such a form `void(0)` is pretty common in js to protect from imaginary undefined reassignment :)

Maybe it worth to add and exception and check just for `void(0)`? Would assume it is most commonly used void type :)

Or there are some technical complication in doing that?

**Comment by ajafff** [comment]
> Maybe it worth to add and exception and check just for void(0)? Would assume it is most commonly used void type :)

I thought `void 0` was most commonly used when I made that change back then. I'm totally fine with adding another exception for `void(0)`.
